### ISSUE TITLE
Pygmentize ansi escape sequences are stripped from Windows pipes

### ISSUE THREAD

**Issue by rmccampbell** [pygments/pygments#2491]
Pygmentize wraps stdout with colorama on Windows. This ensures correct color output on the console but when output is redirected to a pipe it strips the ansi sequences by default. This prevents me from piping the output to `less` or other tools. On the modern windows terminal ansi escapes are handled directly by the terminal anyway so colorama is not necessary, and even on older windows tools like Cygwin's less/cat handle ansi sequences internally so being able to pipe them is still useful.

The fix I think is either setting `strip=False` in colorama.init() so it won't touch non-console output, or (for the most recent colorama version) calling `colorama.just_fix_windows_console` instead (see https://github.com/tartley/colorama/pull/139#issuecomment-1278346238)

**Comment by jofish920** [comment]
What about adding an option to `pygments/cmdline.py` that disables all of the special handling for Windows Terminal?  This is what I think was suggested as the preferred option at the end of the discussion for issue #2492.  Adding an option "--color" that causes the special handling for Windows to be disabled is an easy fix.  I didn't try to do anything fancy -- just added an option and a test that the option was not set to the checks that the script is running on Windows, using a terminal formatter, and not sending output to a named file.  I encountered a few issues in using ``pytest`` to test the changes, which work when the package is installed, but I can work those out and submit a pull request if anyone is interested?

This was mostly done to address my own irritation when using ``pygmentize`` under Windows, but I can't see and major issues with the solution.  It only affects calls to ``main`` in which the previously unused "--color" option is provided, and does nothing more than disable some special handling under Windows.  The semantics could be extended to handle similar cases with other systems, but I don't think there are any right now.

**Comment by jofish920** [comment]
The pytest issue was that I forgot to set the formatter to "terminal" in the command line test; I fixed it, and the test now passes.  The tox tests too, though that was expected.  A total of 18 lines added and 1 slightly changed, including the unit test, which I added to `tests/test_cmdline.py`.

**Comment by rmccampbell** [comment]
The resolution in that PR was not to disable all Windows handling, but just the escape stripping when redirecting to a file/pipe, and to have the stripping opt-in and available on all platforms.

**Comment by rmccampbell** [comment]
(I believe that the ansi-to-win32 conversion behavior will be automatically disabled by colorama when it detects the windows terminal's built in ansi support anyway. But it still enables the escape stripping when isatty is false, with the current arguments.)

**Comment by jofish920** [comment]
Thanks for the clarification.  I appreciate you getting back to me so quickly.

I skimmed the discussion, and guess I got out of it what I wanted to read :(
I've used the tool on other systems, and finding things not working under
Windows was irritating.  The easy fix was a bit too appealing, I guess.
Despite this, the idea of being able to override settings that are normally
inferred from the environment does still appeal to me.

What would you like me to do with this?  I'm not hugely invested in the fix;
it was just the product of fixing something that was annoying me.
Commenting on it has taken longer than making the changes.

On the other hand, if you would like me to put together something
more in line with what was suggested, it doesn't sound like a
bad idea.  It would be nice if there was a way of resolving problems
with piping.

**Comment by rmccampbell** [comment]
So my fix in PR #2492 solved the issue just fine but I think the project owners want to be able to re-enable that behavior by flag, and also have the flag apply on Unix as well. Experimentally it seems that colorama can wrap a stream and strip escapes even on Linux so that shouldn't be a hard change, but Anteru implied they were going to make the change themself so I didn't bother following up on it.

**Comment by jofish920** [comment]
Fair enough; your change seems to give what I originally wanted as the default behaviour.  It has been more than a year since you proposed the change, though, and the anticipated change does not seem to have made it into the main branch of the code.  I have altered my code so that it extends what I believe (on re-reading) was suggested in the comments on PR #2492.  What I understood of the officially proposed solution is that it would introduce an option "--strip-escapes" that can have values "on", "off", and "auto", which will produce behaviours of stripping escapes on all platforms, leaving escapes in place (while converting them to work in the detected environment), and applying the current heuristics to determine what should be done.   I would still like to support the option of disabling the post-processing, just in case something else goes wrong, and have done this by "raw" as a possible value for "--strip-escapes". 

Currently the tests for "--strip-escapes=off" fail, while those for "--strip-escapes=raw" succeed.  This is odd, since running the script with the two options produces identical output but for the conversion of '\n' to '\r\n' in the output with "--strip-escapes=off".  I checked by running the two under Windows, with output redirected to file, used `sed` under `wsl` to translate the line endings for the "raw" file and put both files through `od -a` before diffing them (for readability).  No differences.  Under `pytest`, the "off" and "semi-raw" cases cut off at around 32808 characters.  I suspect a bug in the test, but it isn't obvious, and I don't currently have time to check.

I'm going to push the changes to my repo anyway.  If you feel like looking, you are welcome; if not, this is mostly for my own satisfaction at this point.  Changes are up to about 70 lines, including tests and making '--color' an alias for '--strip-escapes=raw'.

**Comment by rmccampbell** [comment]
I would just send a PR.

**Comment by jofish920** [comment]
Done.  PR #2802.

**Comment by jofish920** [comment]
I've looked into the test failure that I mentioned earlier, and it looks like it has something to do with a file not being flushed in `pygments.terminal.format_unencoded`.  Tracing through with a debugger revealed that the underlying `BytesIO` object is not updated by the time the function exits, and calling `flush` from the debugger at that point causes the test to pass.  One is called as a file-like object is closed shortly thereafter, but that does not appear to fix things.  Maybe a different object?  I skipped over a few steps in tracing.

I suspect that I should raise this as a separate issue and add the `flush` in `format_unencoded`.  Determining exactly where the data gets lost is still something that needs to be done, but flushing the output at the end of this function makes sense.

**Comment by jofish920** [comment]
After writing a nice test case designed to demonstrate the flushing problem independantly from the fix to this issue, I realised that unit-testing using the terminal would trigger the original bug,  This would cause the output to be stripped of escape sequences, which means that the bug does not trigger, at least with the input file that I was using.  I knew this at the start, because the issue with the test case for PR #2802 only surfaces with "--strip-escapes=off" or "--strip-escapes=semi-raw", which are the two cases where the colorama wrapper is used but the output is not stripped.

I am adding the fix to my PR until I can reproduce the bug without my fix to this issue, at which point I will open another issue.

**Comment by jofish920** [comment]
Amend the "with the input file I was using" to "with any file in the test directory".  The difference in behaviour does not seem dependant on the number of bytes written to outfile.  I'll track down the exact source at a later date; for now, I have a fix to apply.

**Comment by birkenfeld** [comment]
I wonder if it's not easiest to use `colorama.just_fix_windows_console()`. I overlooked that solution initially.

**Comment by rmccampbell** [comment]
Yes, I also think that is the simplest solution which I mentioned in the original post, assuming that we don't care about tying to a newer version of colorama. But from the discussion in the closed PR they were reluctant to make that change without adding an option to enable stripping in a cross-platform way.

Personally I think that could be left to a separate issue, because I think the number of people relying on the current stripping feature on pipes/redirection on windows are likely very small, and it was already agreed that changing the default behavior is acceptable here.

**Comment by rmccampbell** [comment]
Though looking at this again, the just_fix_windows_console approach is a bit nuanced bc it only wraps the text streams (on old windows where the wrapper is still needed) but not the underlying binary streams, and currently pygments will re-wrap the windows binary stdout stream in a new text stream and the colorama wrapper. So to more closely preserve that behavior I think calling wrap_stream manually still may be needed.

Also just_fix_windows_console appears to have a subtle bug which may not matter much here but could break some usages of stdout/stderr, which I just filed a PR to fix: https://github.com/tartley/colorama/pull/397

**Comment by birkenfeld** [comment]
Ok, even simpler - since Python 3.9 is the oldest non-EOL version, and it doesn't run on Windows 7: do the nuanced behavior and the subtle bug still matter? :) AIUI, there is no wrapping at all in Windows 10+, only the console "ioctl" equivalent to switch on processing.

**Comment by rmccampbell** [comment]
Windows 8?

**Comment by birkenfeld** [comment]
Probably not. And Windows 8 must be a myth anyway 🤣
